File: desktop/src/shared/version-compare.ts
/**
 * Phase 3b: semver-ish comparison. Returns true if `latest` is a greater
 * version than `installed`. Falls back to strict inequality for non-numeric
 * version strings (different = update available).
 */
export function isNewerVersion(installed: string | undefined, latest: string | undefined): boolean {
  if (!installed || !latest) return false;
  const stripV = (v: string) => v.replace(/^v/i, '').trim();
  const a = stripV(installed);
  const b = stripV(latest);
  if (a === b) return false;
  const parse = (v: string) => v.split(/[.\-+]/).map(p => /^\d+$/.test(p) ? parseInt(p, 10) : NaN);
  const pa = parse(a);
  const pb = parse(b);
  // If any segment is NaN, fall back to string inequality (different = update)
  if (pa.some(isNaN) || pb.some(isNaN)) return a !== b;
  const len = Math.max(pa.length, pb.length);
  for (let i = 0; i < len; i++) {
    const ai = pa[i] ?? 0;
    const bi = pb[i] ?? 0;
    if (bi > ai) return true;
    if (bi < ai) return false;
  }
  return false;
}

File: app/src/main/kotlin/com/youcoded/app/skills/VersionCompare.kt
package com.youcoded.app.skills

/**
 * Port of desktop's `desktop/src/shared/version-compare.ts` (`isNewerVersion`)
 * — keep the two in step. Returns true if `latest` is a greater dotted
 * version than `installed`. Falls back to strict string inequality for
 * non-numeric version strings (different string = treat as an update).
 */
object VersionCompare {
    fun isNewer(installed: String?, latest: String?): Boolean {
        if (installed.isNullOrBlank() || latest.isNullOrBlank()) return false
        // Strip one leading "v"/"V", same as the TS `replace(/^v/i, '')`.
        val a = installed.removePrefix("v").removePrefix("V").trim()
        val b = latest.removePrefix("v").removePrefix("V").trim()
        if (a == b) return false
        val pa = a.split('.', '-', '+').map { it.toIntOrNull() }
        val pb = b.split('.', '-', '+').map { it.toIntOrNull() }
        // Any non-numeric segment on either side → fall back to "different = update".
        if (pa.any { it == null } || pb.any { it == null }) return a != b
        for (i in 0 until maxOf(pa.size, pb.size)) {
            val ai = pa.getOrNull(i) ?: 0
            val bi = pb.getOrNull(i) ?: 0
            if (bi > ai) return true
            if (bi < ai) return false
        }
        return false
    }
}
